### PR TITLE
chore: set up pre-commit and Makefile for local development

### PR DESCRIPTION
**Pull request by thompsonson**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [X] locally yes

# Suggested Checklist:

- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] My changes generate no new warnings
- [ ] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods
- [X] `optional` I have made corresponding changes to the documentation 
- [ ] `optional` I have added tests that prove my fix is effective or that my feature works
- [ ] `optional` New and existing unit tests pass locally with my changes

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,29 @@
+exclude: '^docs/|/migrations/|devcontainer.json'
+default_stages: [commit]
+
+default_language_version:
+  python: python3.11
```
Comment (bowenliang123) [inline .pre-commit-config.yaml]: python 3.10

https://docs.dify.ai/getting-started/install-self-hosted/local-source-code#installation-of-the-basic-environment

File: dev/README_CN.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+### 单元测试
+
+评估工具功能：
+
+```sh
+pytest api/tests/unit_tests/
+```
+
+通过DeepL.com（免费版）翻译
```
Comment (bowenliang123) [inline dev/README_CN.md]: Remove all the footers for translation tool.

File: dev/README_ES.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+### Pruebas unitarias
+
+Evaluar la funcionalidad de la herramienta:
+
+```sh
+pytest api/tests/unit_tests/
+```
+
+Traducción realizada con la versión gratuita del traductor DeepL.com
```
Comment (bowenliang123) [inline dev/README_ES.md]: ditto

File: dev/README_FR.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+### Tests unitaires
+
+Évaluer la fonctionnalité de l'outil :
+
+``sh
+pytest api/tests/unit_tests/
+```
+
+Traduit avec DeepL.com (version gratuite)
```
Comment (bowenliang123) [inline dev/README_FR.md]: ditto

File: dev/README_JA.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+pytest api/tests/integration_tests/ を実行します。
+```
+
+### ユニットテスト
+
+ツールの機能を評価する：
+
+以下のようにします。
+pytest api/tests/unit_tests/ ### ユニットテスト
+```
```
Comment (bowenliang123) [inline dev/README_JA.md:55]: ditto

File: dev/pre-commit [hunk 6]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env bash
+# File generated by pre-commit: https://pre-commit.com
+# ID: 138fd403232d2ddd5efb44317e38bf03
+
+# start templated
+INSTALL_PYTHON=/usr/bin/python3
```
Comment (bowenliang123) [inline dev/pre-commit]: do not use hardcoded absolute path for Python. Developers often relies on python env tools like venv and conda, which lead to relative python path.

File: dev/pre-commit [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+HERE="$(cd "$(dirname "$0")" && pwd)"
+ARGS+=(--hook-dir "$HERE" -- "$@")
+
+if [ -x "$INSTALL_PYTHON" ]; then
+    exec "$INSTALL_PYTHON" -mpre_commit "${ARGS[@]}"
+elif command -v pre-commit > /dev/null; then
+    exec pre-commit "${ARGS[@]}"
+else
+    echo '`pre-commit` not found.  Did you forget to activate your virtualenv?' 1>&2
```
Comment (bowenliang123) [inline dev/pre-commit]:
```suggestion
    echo '`pre-commit` not found.  Did you forget to activate your python environment?' 1>&2
```
Avoid mentioning `venv`. `conda` is suggested to use in doc https://docs.dify.ai/getting-started/install-self-hosted/local-source-code#server-deployment .

File: dev/pre-commit [hunk 8]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env bash
```
Comment (bowenliang123) [inline dev/pre-commit:1]: run` chmod +x dev/pre-commit` to make it executable.

File: dev/pre-commit [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+# File generated by pre-commit: https://pre-commit.com
+# ID: 138fd403232d2ddd5efb44317e38bf03
```
Comment (bowenliang123) [inline dev/pre-commit]: Remove these unhelpful hints. We don't have to rely on this is to modify and regenerate this file.

File: Makefile [hunk 10]
```diff
@@ -0,0 +1,11 @@
+.PHONY: style test install_local_dev
+
+style:
+	ruff api/ --fix
```
Comment (bowenliang123) [inline Makefile]:
```suggestion
	ruff check --fix ./api
```

Use ruff check for calling the linter properly. (ruff contains linter and formtater.)
And ruff usage as below: (from `ruff -h`)
`Usage: ruff [OPTIONS] <COMMAND>`

**Review by bowenliang123 — COMMENTED** [review]
Some comments to follow.
And it does checks with `ruff` on every commit, unlike what we've done in `/web/.husky/pre-commit` for on-demand linting on API and web modules.
 cc @takatost

**Comment by thompsonson** [comment]
> Some comments to follow. And it does checks with `ruff` on every commit, unlike what we've done in `/web/.husky/pre-commit` for on-demand linting on API and web modules. cc @takatost

It looks like there may be some overlap. 

I've made the changed requested below, mindful that Husky might be chosen over pre-commit. 

With either tool I think the Makefile will be useful for new and old developers.

File: dev/pre-commit [hunk 11]
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env bash
+
+# start templated
+INSTALL_PYTHON=$(which python3 || command -v python3)
```
Comment (bowenliang123) [inline dev/pre-commit]:
```suggestion
INSTALL_PYTHON=$(which python || command -v python)
```
`python` should be enough.

Comment (bowenliang123) [reply dev/pre-commit]: The `pyproject.yaml` already requires targted python version > 3.10.

**Comment by takatost** [comment]
Thank you very much for your contribution! 🙏

The rules here are only related to the contents of the `api` directory. `api` and `web` belong to different technology stacks, so the corresponding lint, auto-fix rules, and scripts are also different. Have you considered moving them to the `api` directory?
Also, I suggest considering pre-commit separately for `api` and `web`, as they may have different local environments.

**Comment by thompsonson** [comment]
> Thank you very much for your contribution! 🙏
> 
> The rules here are only related to the contents of the `api` directory. `api` and `web` belong to different technology stacks, so the corresponding lint, auto-fix rules, and scripts are also different. Have you considered moving them to the `api` directory? Also, I suggest considering pre-commit separately for `api` and `web`, as they may have different local environments.


I am starting on the Web development now 😃

My thinking has been to make small changes, better to connect. 

As a development user, I think a central Makefile and consistent approach to pre-commit is best. Sets the right tone and limits DevOps tool sprawl. 

As I am starting on the Web development, I'll learn about Husky, first up does it work for Python code?

If it does I suggest the project continues with just the one pre-commit solution.

**Comment by takatost** [comment]
> > Thank you very much for your contribution! 🙏
> > The rules here are only related to the contents of the `api` directory. `api` and `web` belong to different technology stacks, so the corresponding lint, auto-fix rules, and scripts are also different. Have you considered moving them to the `api` directory? Also, I suggest considering pre-commit separately for `api` and `web`, as they may have different local environments.
> 
> I am starting on the Web development now 😃
> 
> My thinking has been to make small changes, better to connect.
> 
> As a development user, I think a central Makefile and consistent approach to pre-commit is best. Sets the right tone and limits DevOps tool sprawl.
> 
> As I am starting on the Web development, I'll learn about Husky, first up does it work for Python code?
> 
> If it does I suggest the project continues with just the one pre-commit solution.

Yep, having a makefile as an entry point makes a lot of sense. What I mean is, if I only want to make changes to the backend code (Python) of the `api`, then I just need to run the lint rules related to the `api`, without worrying about the dependencies needed for the `web` frontend, and vice versa.

Also, `Husky` is a NodeJS package, so it might not be very suitable for use with Python.

**Comment by thompsonson** [comment]
> What I mean is, if I only want to make changes to the backend code
(Python) of the api, then I just need to run the lint rules related to the
api, without worrying about the dependencies needed for the web frontend,
and vice versa

100%. One of the good things about `pre-commit`is that it will only take
action on the files that are in the git commit.

In terms of moving the files, I'll check if the .pre-commit.yaml can run in
a sub-folder of the git repo.

Would you agree that the Makefile can be used for both `web` and `api` so
is best stored in the root?

To give an example, locally I've added the serve_web PHONY:

.PHONY: style test install_local_dev serve_web

style:
    ruff check --fix ./api

test:
    pytest api/tests

install_local_dev:
    pip install pre-commit ruff
    pre-commit install

serve_web:
    cd web && npm run start

**Comment by takatost** [comment]
> Would you agree that the Makefile can be used for both `web` and `api` so is best stored in the root?

Yeah, putting the Makefile in the root directory makes sense. However, I think it would be a good idea to have separate Makefiles in the `web` and `api` directories as well. The Makefile in the root directory can serve as a central entry point, allowing us to run lint and test uniformly, and it can also execute the Makefiles in the subdirectories.

For example: (root Makefile)

```
.PHONY: lint test lint-web lint-api test-web test-api

# Global lint command that lints both web and API subprojects
lint: lint-web lint-api

# Global test command that tests both web and API subprojects
test: test-web test-api

# Lint command for the web subproject
lint-web:
	$(MAKE) -C web/ lint  # Navigate to web/ directory and run its lint command

# Lint command for the API subproject
lint-api:
	$(MAKE) -C api/ lint  # Navigate to api/ directory and run its lint command

# Test command for the web subproject
test-web:
	$(MAKE) -C web/ test  # Navigate to web/ directory and run its test command

# Test command for the API subproject
test-api:
	$(MAKE) -C api/ test  # Navigate to api/ directory and run its test command
```

**Comment by crazywoola** [comment]
Close as stale.

**Pull request closed without merging**